File: pi-daemon/sdk/send_result_example.c
/*
 * Vi du toi thieu bang C: gui 1 ket qua AI toi carlogd qua UDP noi bo.
 * Bien dich: gcc -O2 -o send_result_example send_result_example.c
 * Copy ham send_ai_result() vao vong lap lai xe cua ban.
 */
#include <stdio.h>
#include <string.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <unistd.h>

static void send_ai_result(double speed_kmh, double steering_deg, double lane_offset_cm,
                            int obstacle, double confidence)
{
    static int sock = -1;
    static struct sockaddr_in addr;
    if (sock < 0) {
        sock = socket(AF_INET, SOCK_DGRAM, 0);
        memset(&addr, 0, sizeof(addr));
        addr.sin_family = AF_INET;
        addr.sin_port = htons(8765); // doi port neu BTC thong bao khac
        inet_pton(AF_INET, "127.0.0.1", &addr.sin_addr);
    }

    char buf[256];
    int len = snprintf(buf, sizeof(buf),
        "{\"speed_kmh\":%.2f,\"steering_deg\":%.2f,\"lane_offset_cm\":%.2f,"
        "\"obstacle\":%s,\"confidence\":%.3f}",
        speed_kmh, steering_deg, lane_offset_cm, obstacle ? "true" : "false", confidence);
    sendto(sock, buf, len, 0, (struct sockaddr *)&addr, sizeof(addr));
}

int main(void)
{
    send_ai_result(25.0, -8.5, 3.2, 0, 0.92);
    printf("da gui 1 goi ket qua AI mau toi carlogd.\n");
    return 0;
}

File: pi-daemon/sdk/send_result_example.py
"""
Vi du toi thieu: gui 1 ket qua AI toi carlogd. Copy ham send_ai_result() vao
vong lap lai xe cua ban, goi no moi khi co ket qua suy luan moi.
"""
import json
import socket

_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
_CARLOGD_ADDR = ("127.0.0.1", 8765)  # doi port neu BTC thong bao khac


def send_ai_result(speed_kmh, steering_deg, lane_offset_cm, obstacle, confidence, progress=None):
    payload = {
        "speed_kmh": speed_kmh,
        "steering_deg": steering_deg,
        "lane_offset_cm": lane_offset_cm,
        "obstacle": obstacle,
        "confidence": confidence,
    }
    if progress is not None:
        payload["progress"] = progress
    _sock.sendto(json.dumps(payload).encode("utf-8"), _CARLOGD_ADDR)


if __name__ == "__main__":
    # Demo: gui thu 1 goi, ban thay bang vong lap lai xe that cua minh.
    send_ai_result(speed_kmh=25.0, steering_deg=-8.5, lane_offset_cm=3.2,
                    obstacle=False, confidence=0.92)
    print("da gui 1 goi ket qua AI mau toi carlogd.")
